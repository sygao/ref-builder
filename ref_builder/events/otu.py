from pydantic import ConfigDict, UUID4, field_serializer

from ref_builder.events.base import ApplicableEvent, EventData, Event, OTUQuery
from ref_builder.resources.otu import RepoOTU
from ref_builder.models import Molecule
from ref_builder.plan import Plan


class CreateOTUData(EventData):
    """The data for a :class:`CreateOTU` event."""

    id: UUID4
    acronym: str
    legacy_id: str | None
    molecule: Molecule
    name: str
    taxid: int
    plan: Plan


class CreateOTU(Event):
    """An event that creates a new OTU."""

    data: CreateOTUData
    query: OTUQuery

    def apply(self) -> RepoOTU:
        return RepoOTU(
            id=self.query.otu_id,
            acronym=self.data.acronym,
            excluded_accessions=set(),
            isolates=[],
            legacy_id=self.data.legacy_id,
            molecule=self.data.molecule,
            name=self.data.name,
            representative_isolate=None,
            plan=self.data.plan,
            taxid=self.data.taxid,
        )


class CreatePlanData(EventData):
    """The data for a :class:`CreatePlan` event."""

    plan: Plan


class CreatePlan(ApplicableEvent):
    """An event that sets the isolate plan for an OTU."""

    data: CreatePlanData
    query: OTUQuery

    def apply(self, otu: RepoOTU) -> RepoOTU:
        """Apply changed plan to OTU and return."""
        otu.plan = self.data.plan

        return otu


class SetRepresentativeIsolateData(EventData):
    """The data for a :class:`SetReprIsolate` event."""

    isolate_id: UUID4


class SetRepresentativeIsolate(ApplicableEvent):
    """An event that sets the representative isolate for an OTU."""

    data: SetRepresentativeIsolateData
    query: OTUQuery

    def apply(self, otu: RepoOTU) -> RepoOTU:
        """Update the OTU's representative isolate and return."""
        otu.representative_isolate = self.data.isolate_id

        return otu


class UpdateExcludedAccessionsData(EventData):
    """The data for an UpdateAllowedAccessions event."""

    model_config = ConfigDict(use_enum_values=True)

    accessions: set[str]
    action: ExcludedAccessionAction

    @field_serializer("accessions")
    def serialize_accessions(self, accessions: set[str]) -> list[str]:
        return sorted(accessions)


class UpdateExcludedAccessions(ApplicableEvent):
    """An event that changes the OTU excluded accessions collection.

    This event is emitted when Genbank accessions are either
    allowed or disallowed from inclusion in the reference.
    """

    data: UpdateExcludedAccessionsData
    query: OTUQuery

    def apply(self, otu: RepoOTU) -> RepoOTU:
        """Add accession allowance changes to OTU and return."""

        if self.data.action == ExcludedAccessionAction.ALLOW:
            for accession in self.data.accessions:
                otu.excluded_accessions.discard(accession)

        else:
            for accession in self.data.accessions:
                otu.excluded_accessions.add(accession)

        return otu
