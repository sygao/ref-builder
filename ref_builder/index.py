"""An index for improving repository performance."""

import binascii
import sqlite3
from collections.abc import Iterator
from dataclasses import dataclass
from pathlib import Path
from typing import Any
from uuid import UUID

import orjson

from ref_builder.models import OTUMinimal
from ref_builder.resources.otu import RepoOTU


@dataclass
class EventIndexItem:
    """The event IDs associated with an OTU."""

    at_event: int
    """The latest event ID when this cache index was last updated"""

    event_ids: list[int]
    """A list of event IDs"""

    otu_id: UUID
    """The OTU ID this cache index is associated with."""


@dataclass
class Snapshot:
    """A snapshot of an OTU at a specific event."""

    at_event: int
    """The event at which the snapshot was taken."""

    otu: RepoOTU
    """The OTU that was snapshotted."""


class Index:
    """An index for rapid access to repository data.

    1. Get OTU IDs by legacy ID, name, or taxonomy ID.
    2. Save and load complete snapshots of OTUs.
    3. Get the event IDS associated with an OTU.

    """

    def __init__(self, path: Path) -> None:
        """Initialize the index."""
        self.path = path
        """The path to the index file."""

        self.path.parent.mkdir(exist_ok=True, parents=True)

        self.con = sqlite3.connect(path, isolation_level=None)
        self.con.execute("PRAGMA journal_mode = WAL")
        self.con.execute("PRAGMA synchronous = NORMAL")

        self.con.execute(
            """
            CREATE TABLE IF NOT EXISTS at_event (
                id INTEGER PRIMARY KEY,
                at_event INTEGER
            );
            """,
        )

        self.con.execute(
            """
            CREATE TABLE IF NOT EXISTS events (
                at_event INTEGER,
                event_id INTEGER PRIMARY KEY,
                otu_id TEXT
            );
            """,
        )

        self.con.execute(
            """
            CREATE TABLE IF NOT EXISTS otus (
                id TEXT PRIMARY KEY,
                acronym TEXT,
                at_event INTEGER,
                legacy_id TEXT,
                name TEXT,
                otu JSONB,
                taxid INTEGER
            );
            """,
        )

        self.con.execute(
            """
            CREATE TABLE IF NOT EXISTS sequences (
                id TEXT PRIMARY KEY,
                crc TEXT,
                otu_id TEXT,
                sequence TEXT
            );
            """,
        )

        for table, column in [
            ("events", "otu_id"),
            ("otus", "id"),
            ("otus", "legacy_id"),
            ("otus", "name"),
            ("otus", "taxid"),
            ("sequences", "otu_id"),
            ("sequences", "crc"),
        ]:
            self.con.execute(
                f"""
                CREATE INDEX IF NOT EXISTS {table}_{column} ON {table} ({column});
                """,
            )

        self.con.commit()

    @property
    def otu_ids(self) -> set[UUID]:
        """A list of all OTUs tracked in the index."""
        return {UUID(row[0]) for row in self.con.execute("SELECT id FROM otus")}

    def add_event_id(self, event_id: int, otu_id: UUID) -> None:
        """Add a new event ID and associated it with a given OTU ID.

        This allows fast lookup of event IDs for a given OTU ID.

        :param event_id: the event ID to add
        :param otu_id: the OTU ID to add the event ID to
        """
        self.con.execute(
            """
            INSERT INTO events (event_id, otu_id)
            VALUES (?, ?)
            """,
            (
                event_id,
                str(otu_id),
            ),
        )

    def get_event_ids_by_otu_id(self, otu_id: UUID) -> EventIndexItem | None:
        """Get a list of event IDs for an OTU."""
        res = self.con.execute(
            """
            SELECT event_id
            FROM events
            WHERE otu_id = ?
            ORDER BY event_id
            """,
            (str(otu_id),),
        )

        event_ids = [row[0] for row in res]

        if event_ids:
            return EventIndexItem(
                event_ids[-1],
                event_ids,
                otu_id,
            )

        return None

    def get_id_by_legacy_id(self, legacy_id: str) -> UUID | None:
        """Get an OTU ID by its legacy ID."""
        cursor = self.con.execute(
            "SELECT id FROM otus WHERE legacy_id = ?",
            (legacy_id,),
        )

        if result := cursor.fetchone():
            return UUID(result[0])

        return None

    def get_id_by_name(self, name: str) -> UUID | None:
        """Get an OTU ID by its name."""
        cursor = self.con.execute(
            "SELECT id FROM otus WHERE name = ?",
            (name,),
        )

        if result := cursor.fetchone():
            return UUID(result[0])

        return None

    def get_id_by_taxid(self, taxid: int) -> UUID | None:
        """Get an OTU ID by its taxonomy ID."""
        cursor = self.con.execute(
            "SELECT id FROM otus WHERE taxid = ?",
            (taxid,),
        )

        if result := cursor.fetchone():
            return UUID(result[0])

        return None

    def delete_otu(self, otu_id: UUID) -> None:
        """Remove an OTU from the index.

        This happens rarely, so we just rewrite the whole index.

        :param otu_id: the ID of the OTU to remove.
        """
        self.con.execute(
            "DELETE FROM events WHERE otu_id = ?",
            (str(otu_id),),
        )

        self.con.execute(
            "DELETE FROM otus WHERE id = ?",
            (str(otu_id),),
        )

        self.con.execute(
            "DELETE FROM sequences WHERE otu_id = ?",
            (str(otu_id),),
        )

        self.con.commit()

    def iter_minimal_otus(self) -> Iterator[OTUMinimal]:
        """Iterate over minimal representations of all OTUs in the index."""
        cursor = self.con.execute(
            "SELECT acronym, id, legacy_id, name, taxid FROM otus ORDER BY name",
        )

        for row in cursor:
            yield OTUMinimal(
                acronym=row[0],
                id=UUID(row[1]),
                legacy_id=row[2],
                name=row[3],
                taxid=row[4],
            )

    def load_snapshot(self, otu_id: UUID) -> Snapshot | None:
        """Load an OTU snapshot."""
        cursor = self.con.execute(
            "SELECT at_event, otu FROM otus WHERE id = ?",
            (str(otu_id),),
        )

        result = cursor.fetchone()

        if not result:
            return None

        at_event, otu_json = result

        otu = orjson.loads(otu_json)

        sequence_ids = [
            str(sequence["id"])
            for isolate in otu["isolates"]
            for sequence in isolate["sequences"]
        ]

        # Fetch all sequences in a single query
        placeholders = ",".join("?" for _ in sequence_ids)

        sequence_map = dict(
            self.con.execute(
                f"SELECT id, sequence FROM sequences WHERE id IN ({placeholders})",
                sequence_ids,
            ).fetchall(),
        )

        # Update the data structure and check for missing sequences
        missing_sequences = []

        for isolate in otu["isolates"]:
            for sequence in isolate["sequences"]:
                sequence_id = str(sequence["id"])

                if sequence_id in sequence_map:
                    sequence["sequence"] = sequence_map[sequence_id]
                else:
                    missing_sequences.append(sequence_id)

        # Raise an error if any sequences are missing
        if missing_sequences:
            raise ValueError(f"Sequences not found: {', '.join(missing_sequences)}")

        return Snapshot(
            at_event=at_event,
            otu=RepoOTU.model_validate(otu),
        )

    def upsert_otu(self, otu: RepoOTU, at_event: int) -> None:
        """Update the index based on a list of OTUs.

        OTUs that aren't already in the index will be added, and those that are will be
        updated.

        Only sequences that have changed will be updated.

        """
        sequence_ids = {
            str(seq.id) for isolate in otu.isolates for seq in isolate.sequences
        }

        otu_dict = otu.model_dump(mode="json")

        placeholders = ",".join("?" for _ in sequence_ids)

        # Delete any sequences that are no longer in the OTU.
        self.con.execute(
            f"""
            DELETE FROM sequences WHERE otu_id = ? AND NOT id IN ({ placeholders });
            """,
            (str(otu.id), *sequence_ids),
        )

        batch = []

        # Insert or update the sequences.
        for isolate in otu_dict["isolates"]:
            for sequence in isolate["sequences"]:
                crc = _calculate_crc32(sequence["sequence"])
                batch.append(
                    (
                        str(sequence["id"]),
                        crc,
                        str(otu.id),
                        sequence["sequence"],
                    ),
                )

        # Perform batch insert
        self.con.executemany(
            """
            INSERT OR REPLACE INTO sequences (id, crc, otu_id, sequence)
            VALUES (?, ?, ?, ?)
            """,
            batch,
        )

        # Update only if CRC has changed
        self.con.executemany(
            """
            UPDATE sequences
            SET crc = ?, otu_id = ?, sequence = ?
            WHERE id = ? AND crc != ?
            """,
            [(crc, otu_id, seq, seq_id, crc) for seq_id, crc, otu_id, seq in batch],
        )

        self.con.execute(
            """
            INSERT OR REPLACE INTO otus (id, acronym, at_event, legacy_id, name, otu, taxid)
            VALUES (?, ?,?, ?, ?, ?, ?)
            """,
            (
                str(otu.id),
                otu.acronym,
                at_event,
                otu.legacy_id,
                otu.name,
                orjson.dumps(otu_dict),
                otu.taxid,
            ),
        )

        self.con.commit()


def _calculate_crc32(sequence: str) -> str:
    """Calculate the CRC32 checksum of a sequence.

    :param sequence: the sequence to calculate the checksum of.
    :return: the CRC32 checksum as a hexadecimal string.

    """
    crc = binascii.crc32(sequence.encode("utf-8"))

    # Convert CRC as a hexadecimal string.
    return hex(crc & 0xFFFFFFFF)[2:].zfill(8)


def _default_json(obj: Any) -> list:
    """Cast sets to lists.

    This function is used as the default argument to `orjson.dumps` to handle sets.

    :param obj: the object to serialize
    :return: the serialized object
    """
    if isinstance(obj, set):
        return list(obj)

    raise TypeError(f"Object of type {obj.__class__.__name__} is not JSON serializable")
