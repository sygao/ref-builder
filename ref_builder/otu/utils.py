import re
from collections import defaultdict
from enum import StrEnum
from uuid import UUID

import structlog

from ref_builder.models import Molecule
from ref_builder.ncbi.models import NCBIGenbank
from ref_builder.plan import (
    MonopartitePlan,
    MultipartitePlan,
    Segment,
    SegmentRule,
)
from ref_builder.utils import Accession, IsolateName, IsolateNameType

logger = structlog.get_logger("otu.utils")


class DeleteRationale(StrEnum):
    """Default strings delineating reasons for resource deletion."""

    USER = "Requested by user"
    REFSEQ = "Superceded by RefSeq"


class RefSeqConflictError(ValueError):
    """Raised when a potential RefSeq replacement is found."""

    def __init__(
        self,
        message: str,
        isolate_id: UUID,
        isolate_name: IsolateName,
        accessions: list[str],
    ):
        super().__init__(message)

        self.isolate_id = isolate_id

        self.isolate_name = isolate_name

        self.accessions = accessions


def create_segments_from_records(
    records: list[NCBIGenbank], rule: SegmentRule, length_tolerance: float
) -> list[Segment]:
    """Return a list of SegmentPlans."""
    segments = []

    for record in sorted(records, key=lambda record: record.accession):
        if not record.source.segment:
            raise ValueError("No segment name found for multipartite OTU segment.")

        segments.append(
            Segment.from_record(
                record=record,
                length_tolerance=length_tolerance,
                required=rule,
            ),
        )

    return segments


def create_isolate_plan_from_records(
    records: list[NCBIGenbank],
    length_tolerance: float,
    segments: list[Segment] | None = None,
) -> MonopartitePlan | MultipartitePlan | None:
    """Return a plan from a list of records representing an isolate."""
    if len(records) == 1:
        return MonopartitePlan.new(
            length=len(records[0].sequence),
            length_tolerance=length_tolerance,
        )

    binned_records = group_genbank_records_by_isolate(records)
    if len(binned_records) > 1:
        logger.fatal(
            "More than one isolate found. Cannot create schema automatically.",
            bins=binned_records,
        )
        return None

    if segments is not None:
        return MultipartitePlan.new(segments=segments)

    segments = create_segments_from_records(
        records,
        rule=SegmentRule.REQUIRED,
        length_tolerance=length_tolerance,
    )

    if segments:
        return MultipartitePlan.new(segments=segments)

    return None


def group_genbank_records_by_isolate(
    records: list[NCBIGenbank],
) -> dict[IsolateName, dict[Accession, NCBIGenbank]]:
    """Indexes Genbank records by isolate name."""
    isolates = defaultdict(dict)

    for record in records:
        try:
            isolate_name = _get_isolate_name(record)

            if isolate_name is None:
                logger.debug(
                    "RefSeq record does not contain sufficient source data "
                    " for automatic inclusion. Add this record manually.",
                    accession=record.accession,
                    definition=record.definition,
                    source_data=record.source,
                )
                continue

            versioned_accession = Accession.from_string(record.accession_version)
            isolates[isolate_name][versioned_accession] = record

        except ValueError:
            continue

    return isolates


def parse_refseq_comment(comment: str) -> tuple[str, str]:
    """Parse a standard RefSeq comment."""
    if not comment:
        raise ValueError("Empty comment")

    refseq_pattern = re.compile(r"^(\w+ REFSEQ): [\w ]+. [\w ]+ ([\w]+).")

    match = refseq_pattern.search(comment)

    if match:
        return match.group(1), match.group(2)

    raise ValueError("Invalid RefSeq comment")


def get_molecule_from_records(records: list[NCBIGenbank]) -> Molecule:
    """Return relevant molecule metadata from one or more records."""
    if not records:
        raise IndexError("No records given")

    rep_record = None
    for record in records:
        if record.refseq:
            rep_record = record
            break
    if rep_record is None:
        rep_record = records[0]

    return Molecule.model_validate(
        {
            "strandedness": rep_record.strandedness.value,
            "type": rep_record.moltype.value,
            "topology": rep_record.topology.value,
        },
    )


def check_sequence_length(sequence: str, segment_length: int, tolerance: float) -> bool:
    if len(sequence) < segment_length * (1.0 - tolerance) or len(
        sequence
    ) > segment_length * (1.0 + tolerance):
        return False

    return True


def _get_isolate_name(record: NCBIGenbank) -> IsolateName | None:
    """Get the isolate name from a Genbank record."""
    if record.source.model_fields_set.intersection(
        {IsolateNameType.ISOLATE, IsolateNameType.STRAIN, IsolateNameType.CLONE},
    ):
        for source_type in IsolateNameType:
            if source_type in record.source.model_fields_set:
                return IsolateName(
                    type=IsolateNameType(source_type),
                    value=record.source.model_dump()[source_type],
                )

    if record.refseq:
        return None

    raise ValueError("Record does not contain sufficient source data for inclusion.")
