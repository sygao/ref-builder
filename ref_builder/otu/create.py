import sys

import structlog

from ref_builder.ncbi.client import NCBIClient
from ref_builder.otu.utils import (
    create_isolate_plan_from_records,
    group_genbank_records_by_isolate,
    get_molecule_from_records,
    parse_refseq_comment,
)
from ref_builder.repo import Repo
from ref_builder.resources import RepoOTU

logger = structlog.get_logger("otu.create")


def create_otu(
    repo: Repo,
    taxid: int,
    accessions: list[str],
    acronym: str,
    ignore_cache: bool = False,
) -> RepoOTU | None:
    """Create a new OTU by taxonomy ID and autogenerate a schema.

    Uses the provided accessions to generate a schema and add a first isolate.

    :param repo: the repository to add the OTU to.
    :param taxid: the taxonomy ID to use.
    :param accessions: a list of accessions to use for the schema.
    :param acronym: an alternative name to use during searches.
    :param ignore_cache: whether to ignore the cache.

    """
    otu_logger = logger.bind(taxid=taxid)

    client = NCBIClient(ignore_cache)

    if repo.get_otu_id_by_taxid(taxid):
        raise ValueError(
            f"Taxonomy ID {taxid} has already been added to this reference.",
        )

    taxonomy = client.fetch_taxonomy_record(taxid)

    if taxonomy is None:
        otu_logger.fatal(f"Could not retrieve {taxid} from NCBI Taxonomy")
        return None

    if not acronym and taxonomy.other_names.acronym:
        acronym = taxonomy.other_names.acronym[0]

    records = client.fetch_genbank_records(accessions)

    if len(records) != len(accessions):
        otu_logger.fatal("Could not retrieve all requested accessions.")
        return None

    binned_records = group_genbank_records_by_isolate(records)

    if len(binned_records) > 1:
        otu_logger.fatal(
            "More than one isolate found. Cannot create schema automatically.",
        )
        return None

    molecule = get_molecule_from_records(records)

    plan = create_isolate_plan_from_records(
        records,
        length_tolerance=repo.settings.default_segment_length_tolerance,
    )

    if plan is None:
        otu_logger.fatal("Could not create plan from records.")
        return None

    try:
        otu = repo.create_otu(
            acronym=acronym,
            legacy_id=None,
            molecule=molecule,
            name=taxonomy.name,
            plan=plan,
            taxid=taxid,
        )
    except ValueError as e:
        otu_logger.fatal(e)
        sys.exit(1)

    isolate = repo.create_isolate(
        otu_id=otu.id,
        legacy_id=None,
        name=next(iter(binned_records.keys())) if binned_records else None,
    )

    otu.add_isolate(isolate)
    otu.repr_isolate = repo.set_repr_isolate(otu_id=otu.id, isolate_id=isolate.id)

    for record in records:
        sequence = repo.create_sequence(
            otu_id=otu.id,
            accession=record.accession_version,
            definition=record.definition,
            legacy_id=None,
            segment=record.source.segment,
            sequence=record.sequence,
        )

        repo.link_sequence(otu.id, isolate.id, sequence.id)

        if record.refseq:
            _, old_accession = parse_refseq_comment(record.comment)
            repo.exclude_accession(
                otu.id,
                old_accession,
            )

    return repo.get_otu(otu.id)
