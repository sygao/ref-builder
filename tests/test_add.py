import subprocess
from pathlib import Path

import pytest
from click.testing import CliRunner
from syrupy import SnapshotAssertion
from syrupy.filters import props

from ref_builder.cli.main import isolate_create, otu_create
from ref_builder.otu.create import create_otu
from ref_builder.otu.update import (
    add_and_name_isolate,
    add_genbank_isolate,
    add_unnamed_isolate,
    update_isolate_from_accessions,
)
from ref_builder.otu.utils import RefSeqConflictError
from ref_builder.repo import Repo
from ref_builder.utils import IsolateName, IsolateNameType


def run_create_otu_command(
    path: Path,
    taxid: int,
    accessions: list,
    acronym: str = "",
):
    subprocess.run(
        [
            "ref-builder",
            "otu",
            "create",
            str(taxid),
            *accessions,
            "--path",
            str(path),
            "--acronym",
            acronym,
        ],
        check=False,
    )


def run_update_otu_command(taxid: int, path: Path):
    subprocess.run(
        ["ref-builder", "otu", "update"] + [str(taxid)] + ["--path", str(path)],
        check=False,
    )


class TestCreateOTU:
    def test_duplicate_accessions(self, precached_repo: Repo):
        """Test that an error is raised when duplicate accessions are provided."""
        runner = CliRunner()
        result = runner.invoke(
            otu_create,
            [
                "--path",
                str(precached_repo.path),
                "1169032",
                "MK431779",
                "MK431779",
            ],
        )

        assert result.exit_code == 2
        assert "Duplicate accessions are not allowed." in result.output

    def test_empty_success(
        self,
        precached_repo: Repo,
        snapshot: SnapshotAssertion,
    ):
        """Test that an OTU can be created in an empty repository."""
        otu = create_otu(
            precached_repo,
            345184,
            ["DQ178610", "DQ178611"],
            "",
        )

        assert otu.model_dump() == snapshot(
            exclude=props("id", "isolates", "repr_isolate"),
        )

        # Ensure only one OTU is present in the repository, and it matches the return
        # value of the creation function.
        assert list(precached_repo.iter_otus()) == [otu]

    def test_empty_fail(self, scratch_repo: Repo):
        with pytest.raises(ValueError):
            create_otu(
                scratch_repo,
                345184,
                ["DQ178610", "DQ178611"],
                "",
            )

    @pytest.mark.parametrize(
        ("taxid", "accessions"),
        [(1278205, ["NC_020160"]), (345184, ["DQ178610", "DQ178611"])],
    )
    def test_otu_create(
        self,
        taxid: int,
        accessions: list[str],
        precached_repo: Repo,
    ):
        assert list(precached_repo.iter_otus()) == []

        otu = create_otu(
            precached_repo,
            taxid,
            accessions,
            "",
        )

        assert list(precached_repo.iter_otus())
        assert otu.plan is not None
        assert otu.repr_isolate is not None

    def test_otu_create_refseq_autoexclude(
        self,
        precached_repo: Repo,
        snapshot: SnapshotAssertion,
    ):
        """Test that the superceded accessions included in RefSeq metadata
        are automatically added to the OTU's excluded accessions list.
        """
        otu = create_otu(
            precached_repo,
            3158377,
            [
                "NC_010314",
                "NC_010316",
                "NC_010315",
                "NC_010317",
                "NC_010318",
                "NC_010319",
            ],
            "",
        )

        assert (
            otu.excluded_accessions
            == {
                "EF546808",
                "EF546809",
                "EF546810",
                "EF546811",
                "EF546812",
                "EF546813",
            }
            == snapshot
        )

    def test_otu_create_with_acronym_auto(self, precached_repo: Repo):
        otu = create_otu(
            precached_repo,
            132477,
            ["NC_013006"],
            "",
        )

        assert otu.acronym == "KLV"

    def test_otu_create_with_acronym_manual(self, precached_repo: Repo):
        otu = create_otu(
            precached_repo,
            1441799,
            ["NC_023881"],
            "FBNSV",
        )

        assert otu.acronym == "FBNSV"


class TestCreateOTUCommands:
    @pytest.mark.parametrize(
        "taxid, accessions",
        [(1278205, ["NC_020160"]), (345184, ["DQ178610", "DQ178611"])],
    )
    def test_ok(
        self,
        taxid: int,
        accessions: list[str],
        precached_repo: Repo,
        snapshot: SnapshotAssertion,
    ):
        run_create_otu_command(
            taxid=taxid,
            path=precached_repo.path,
            accessions=accessions,
        )

        otus = list(Repo(precached_repo.path).iter_otus())

        assert len(otus) == 1
        otu = otus[0]

        assert otu.model_dump() == snapshot(
            exclude=props("id", "isolates", "repr_isolate"),
        )

    def test_acronym(self, precached_repo: Repo):
        """Test if the --acronym option works as planned."""
        run_create_otu_command(
            taxid=345184,
            accessions=["DQ178610", "DQ178611"],
            path=precached_repo.path,
            acronym="CabLCJV",
        )

        otus = list(Repo(precached_repo.path).iter_otus())

        assert len(otus) == 1
        assert otus[0].acronym == "CabLCJV"


class TestAddIsolate:
    def test_duplicate_accessions(self, precached_repo: Repo):
        """Test that an error is raised when duplicate accessions are provided."""
        runner = CliRunner()
        result = runner.invoke(
            isolate_create,
            [
                "345184",
                "DQ178610",
                "DQ178610",
            ],
        )

        assert result.exit_code == 2
        assert "Duplicate accessions are not allowed." in result.output

    def test_genbank_ok(self, precached_repo: Repo):
        """Test that add_genbank_isolate() adds an isolate with a correctly parsed
        name.
        """
        isolate_1_accessions = ["DQ178610", "DQ178611"]
        isolate_2_accessions = ["DQ178613", "DQ178614"]

        otu = create_otu(precached_repo, 345184, isolate_1_accessions, acronym="")

        assert otu.accessions == set(isolate_1_accessions)

        isolate = add_genbank_isolate(precached_repo, otu, isolate_2_accessions)

        otu_after = precached_repo.get_otu_by_taxid(345184)

        assert otu_after.accessions == set(isolate_1_accessions).union(
            set(isolate_2_accessions),
        )

        isolate_after = otu_after.get_isolate(isolate.id)

        assert isolate_after.accessions == set(isolate_2_accessions)

        assert isolate_after.name == IsolateName(
            IsolateNameType.ISOLATE,
            "Douglas Castle",
        )

    def test_ignore_name_ok(self, precached_repo: Repo):
        """Test that add_unnamed_isolate() adds the correct isolate with its name set to None."""
        isolate_1_accessions = ["DQ178610", "DQ178611"]
        isolate_2_accessions = ["DQ178613", "DQ178614"]

        otu = create_otu(precached_repo, 345184, isolate_1_accessions, acronym="")

        assert otu.accessions == set(isolate_1_accessions)

        isolate = add_unnamed_isolate(precached_repo, otu, isolate_2_accessions)

        otu_after = precached_repo.get_otu_by_taxid(345184)

        assert otu_after.isolate_ids == {otu_after.repr_isolate, isolate.id}

        isolate_after = otu_after.get_isolate(isolate.id)

        assert isolate_after.name is None

        assert isolate_after.accessions == {"DQ178613", "DQ178614"}

    def test_add_and_name_ok(self, precached_repo: Repo):
        """Test that add_and_name_isolate() creates an isolate with the correct name."""
        isolate_1_accessions = ["DQ178610", "DQ178611"]
        isolate_2_accessions = ["DQ178613", "DQ178614"]

        otu = create_otu(precached_repo, 345184, isolate_1_accessions, acronym="")

        assert otu.accessions == set(isolate_1_accessions)

        isolate = add_and_name_isolate(
            precached_repo,
            otu,
            isolate_2_accessions,
            isolate_name=IsolateName(type=IsolateNameType.ISOLATE, value="dummy"),
        )

        otu_after = precached_repo.get_otu_by_taxid(345184)

        assert otu_after.isolate_ids == {otu_after.repr_isolate, isolate.id}

        isolate_after = otu_after.get_isolate(isolate.id)

        assert isolate_after.name == IsolateName(
            type=IsolateNameType.ISOLATE,
            value="dummy",
        )

        assert isolate_after.accessions == {"DQ178613", "DQ178614"}

    def test_conflict_fail(self, precached_repo: Repo):
        """Test that an isolate cannot be added to an OTU
        if both its name and its accessions are already contained.
        """
        taxid = 2164102
        accessions = ["MF062136", "MF062137", "MF062138"]

        otu = create_otu(
            precached_repo,
            taxid,
            accessions,
            "",
        )

        assert add_genbank_isolate(precached_repo, otu, accessions) is None


@pytest.mark.parametrize(
    "taxid, original_accessions, refseq_accessions",
    [
        (1169032, ["AB017503"], ["NC_003355"]),
        (345184, ["DQ178608", "DQ178609"], ["NC_038792", "NC_038793"]),
    ],
)
class TestReplaceIsolateSequences:
    def test_manual_replace_ok(
        self,
        empty_repo,
        taxid,
        original_accessions,
        refseq_accessions,
    ):
        """Test that a requested replacement occurs as expected."""
        create_otu(
            empty_repo,
            taxid,
            accessions=original_accessions,
            acronym="",
        )

        otu_before = empty_repo.get_otu_by_taxid(taxid)

        assert otu_before.accessions == set(original_accessions)

        isolate = list(otu_before.isolates)[0]

        update_isolate_from_accessions(
            empty_repo,
            otu_before,
            isolate.name,
            refseq_accessions,
        )

        otu_after = empty_repo.get_otu(otu_before.id)

        assert otu_after.accessions == set(refseq_accessions)

        assert otu_after.excluded_accessions == set(original_accessions)

    def test_raise_refseq_exception(
        self,
        empty_repo,
        taxid: int,
        original_accessions: list[str],
        refseq_accessions: list[str],
    ):
        """Test that attempting to add an isolate with RefSeq accessions
        raises RefSeqConflictError
        """
        create_otu(
            empty_repo,
            taxid,
            accessions=original_accessions,
            acronym="",
        )

        otu_before = empty_repo.get_otu_by_taxid(taxid)

        assert otu_before.accessions == set(original_accessions)

        add_genbank_isolate(empty_repo, otu_before, original_accessions)

        with pytest.raises(RefSeqConflictError):
            add_genbank_isolate(empty_repo, otu_before, refseq_accessions)
