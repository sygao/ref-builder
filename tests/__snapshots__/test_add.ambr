# serializer version: 1
# name: TestCreateOTU.test_empty_success
  dict({
    'acronym': '',
    'excluded_accessions': set({
    }),
    'legacy_id': None,
    'molecule': dict({
      'strandedness': 'double',
      'topology': 'circular',
      'type': 'DNA',
    }),
    'name': 'Cabbage leaf curl Jamaica virus',
    'plan': dict({
      'plan_type': 'multipartite',
      'segments': list([
        dict({
          'length': 2575,
          'length_tolerance': 0.03,
          'name': dict({
            'key': 'A',
            'prefix': 'DNA',
          }),
          'required': 'required',
        }),
        dict({
          'length': 2494,
          'length_tolerance': 0.03,
          'name': dict({
            'key': 'B',
            'prefix': 'DNA',
          }),
          'required': 'required',
        }),
      ]),
    }),
    'taxid': 345184,
  })
# ---
# name: TestCreateOTU.test_otu_create_refseq_autoexclude
  set({
    'EF546808',
    'EF546809',
    'EF546810',
    'EF546811',
    'EF546812',
    'EF546813',
  })
# ---
# name: TestCreateOTUCommands.test_ok[1278205-accessions0]
  dict({
    'acronym': '',
    'excluded_accessions': set({
      'JX263426',
    }),
    'legacy_id': None,
    'molecule': dict({
      'strandedness': 'single',
      'topology': 'linear',
      'type': 'RNA',
    }),
    'name': 'Dahlia latent viroid',
    'plan': dict({
      'length': 342,
      'length_tolerance': 0.03,
      'name': None,
      'plan_type': 'monopartite',
    }),
    'taxid': 1278205,
  })
# ---
# name: TestCreateOTUCommands.test_ok[345184-accessions1]
  dict({
    'acronym': '',
    'excluded_accessions': set({
    }),
    'legacy_id': None,
    'molecule': dict({
      'strandedness': 'double',
      'topology': 'circular',
      'type': 'DNA',
    }),
    'name': 'Cabbage leaf curl Jamaica virus',
    'plan': dict({
      'plan_type': 'multipartite',
      'segments': list([
        dict({
          'length': 2575,
          'length_tolerance': 0.03,
          'name': dict({
            'key': 'A',
            'prefix': 'DNA',
          }),
          'required': 'required',
        }),
        dict({
          'length': 2494,
          'length_tolerance': 0.03,
          'name': dict({
            'key': 'B',
            'prefix': 'DNA',
          }),
          'required': 'required',
        }),
      ]),
    }),
    'taxid': 345184,
  })
# ---
